### PR TITLE
[CARBONDATA-3986] Fix multiple issues during compaction and concurrent scenarios

### PR DESCRIPTION
**Pull request by ajantha-bhat**
### Why is this PR needed?
Fix multiple issues during compaction and concurrent scenarios
a)  Auto compaction/multiple times minor compaction is called, it was considering compacted segments and coming compaction again ad overwriting the files and segments
b) Minor/ auto compaction should skip >=2 level segments, now only skipping =2 level segments
c) when compaction failed, no need to call merge index
d) At executor, When segment file or table status file failed to write during merge index event, need to remove the stale files.
e) during partial load cleanup segment folders are removed but segment metadata files were not removed
f) Some table status retry issues
g) BlockIndex was not checking for empty partition

 ### What changes were proposed in this PR?
a) Auto compaction/minor compaction consider only SUCCESS or PARTIAL_SUCCESS Segments
b) Minor/ auto compaction should skip >=2 level segments
c) when compaction failed, no need to call merge index
d) At executor, When segment file or table status file failed to write during merge index event, Remove the stale files.
e) During partial load cleanup segment metadata files also need to be removed
f)  add retry to table status 
g) BlockIndex need to check for empty partition     
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - No [as mostly concurrent scenarios]

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3535/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1796/

**Comment by akashrn5** [comment]
@ajantha-bhat please resolve conflicts

File: core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java [hunk 1]
```diff
@@ -234,13 +235,15 @@ public String writeMergeIndexFileBasedOnSegmentFile(String segmentId,
         }
         if (FileFactory.getCarbonFile(entry.getKey()).equals(FileFactory.getCarbonFile(location))) {
           segment.getValue().setMergeFileName(mergeIndexFile);
+          mergeIndexFiles.add(entry.getKey() + "/" + mergeIndexFile);
```
Comment (akashrn5) [inline core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java:238]: use `FILE_SEPARATOR` constant

File: core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java [hunk 2]
```diff
@@ -234,13 +235,15 @@ public String writeMergeIndexFileBasedOnSegmentFile(String segmentId,
...
           segment.getValue().setFiles(new HashSet<String>());
           break;
         }
       }
       if (table.isHivePartitionTable()) {
         for (PartitionSpec partitionSpec : partitionSpecs) {
           if (partitionSpec.getLocation().toString().equals(partitionPath)) {
+            // TODO: handle failure here
```
Comment (akashrn5) [inline core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java:246]: is TODO needs to be handled in this PR? please handle if yes

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonTableCompactor.scala [hunk 3]
```diff
@@ -373,11 +381,7 @@ class CarbonTableCompactor(carbonLoadModel: CarbonLoadModel,
       } else {
         true
       }
-      // here either of the conditions can be true, when delete segment is fired after compaction
-      // has started, statusFileUpdation will be false , but at the same time commitComplete can be
```
Comment (akashrn5) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonTableCompactor.scala:377]: here may be instead of deleting the complete comment, you can rewrite the comment required for `commitComplete`

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala [hunk 4]
```diff
@@ -235,18 +235,6 @@ private[sql] case class CarbonCreateSecondaryIndexCommand(
         throw new ErrorMessage(s"Number of columns in Index table cannot be more than " +
                                "number of key columns in Source table")
       }
-
-      var isColsIndexedAsPerTable = true
```
Comment (akashrn5) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/command/SICreationCommand.scala:239]: can you please tell me why this change required?

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/load/CarbonInternalLoaderUtil.java [hunk 5]
```diff
@@ -173,11 +180,15 @@ public static boolean recordLoadMetadata(List<LoadMetadataDetails> newLoadMetada
...
                 + tableName);
+        throw new RuntimeException(
+            "Not able to acquire the lock for Table status updation for table " + databaseName + "."
+                + tableName);
       }
     } catch (IOException e) {
       LOGGER.error(
           "Not able to acquire the lock for Table status updation for table " + databaseName + "."
               + tableName);
+
```
Comment (akashrn5) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/load/CarbonInternalLoaderUtil.java:191]: revert this change

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/rdd/SecondaryIndexCreator.scala [hunk 6]
```diff
@@ -357,7 +357,6 @@ object SecondaryIndexCreator {
       if (null != executorService) {
         executorService.shutdownNow()
       }
-
```
Comment (akashrn5) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/rdd/SecondaryIndexCreator.scala:360]: revert this change

File: processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java [hunk 7]
```diff
@@ -54,39 +59,77 @@
...
-          }
-        } else {
-          if (!segmentId.contains(".")) {
-            CarbonLoaderUtil.deleteStorePath(listFiles[k].getAbsolutePath());
-          }
+          LOGGER.error(
+              "Not able to acquire the Table status lock for partial load deletion for table "
+                  + carbonTable.getDatabaseName() + "." + carbonTable.getTableName()
+                  + ", retry compaction");
+          throw new RuntimeException(
```
Comment (akashrn5) [inline processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java:125]: Do we really need to throw exception here? because since it's clean up, we can just give an error message and retry next time instead of failing the operation, what do you think?

**Comment by akashrn5** [comment]
@ajantha-bhat please create a JIRA and update the PR description

**Comment by ajantha-bhat** [comment]
@akashrn5 : This is WIP PR, please wait

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4057/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2319/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2327/

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonTableCompactor.scala (hunk 3 above)
Comment (ajantha-bhat) [reply integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonTableCompactor.scala]: the existing comment in the code was that, why we need to consider both the flag. Now only one flag is there. so, it is understandable. 

Based on new code if it is not understandable and if you have any suggestion for comment, you can give here. I will add.

File: processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java (hunk 7 above)
Comment (ajantha-bhat) [reply processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java]: yes, for compaction flow I faced the issue. If the stale segments not cleared. there will be duplicate data after compaction. so, I cannot go ahead for compaction if the stale data is not cleaned. so, throwing the exception here.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4066/

**Comment by ajantha-bhat** [comment]
@akashrn5 , It was a WIP PR. Now I got some time and worked on it.
PR is ready you can freshly review and merge.

File: processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java [hunk 8]
```diff
@@ -64,45 +68,89 @@ public static void deletePartialLoadDataIfExist(CarbonTable carbonTable,
...
+          Set<String> staleSegmentsId = new HashSet<>(allSegments.length);
+          for (CarbonFile segment : allSegments) {
+            String segmentName = segment.getName();
+            // check segment folder pattern
+            if (segmentName.startsWith(CarbonTablePath.SEGMENT_PREFIX)) {
+              String[] parts = segmentName.split(CarbonCommonConstants.UNDERSCORE);
+              if (parts.length == 2) {
+                boolean isOriginal = !parts[1].contains(".");
+                if (isCompactionFlow) {
+                  // in compaction flow, it should be big segment and segment metadata is not exists
```
Comment (akashrn5) [inline processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java:101]:
```suggestion
                  // in compaction flow, it should be big segment and segment metadata does not exists
```

File: processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java [hunk 9]
```diff
@@ -64,45 +68,89 @@ public static void deletePartialLoadDataIfExist(CarbonTable carbonTable,
...
+                  if (!isOriginal && !metadataSet.contains(parts[1])) {
+                    staleSegments.add(segment);
+                    staleSegmentsId.add(parts[1]);
+                  }
+                } else {
+                  // in loading flow,
+                  // it should be original segment and segment metadata is not exists
```
Comment (akashrn5) [inline processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java:108]:
```suggestion
                  // it should be original segment and segment metadata does not exists
```

File: processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java [hunk 10]
```diff
@@ -64,45 +68,89 @@ public static void deletePartialLoadDataIfExist(CarbonTable carbonTable,
...
+              CarbonUtil.deleteFoldersAndFiles(staleSegment);
+            } catch (IOException | InterruptedException e) {
+              LOGGER.error("Unable to delete the given path :: " + e.getMessage(), e);
+            }
+          }
+          if (staleSegments.size() > 0) {
+            // collect the segment metadata path
+            String segmentFilesLocation =
+                CarbonTablePath.getSegmentFilesLocation(carbonTable.getTablePath());
+            CarbonFile[] allSegmentMetadataFiles =
```
Comment (akashrn5) [inline processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java:129]: instead of list files all, you can list files whole names starts with id present in `staleSegments` and directly delete those

File: processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java [hunk 11]
```diff
@@ -64,45 +68,89 @@ public static void deletePartialLoadDataIfExist(CarbonTable carbonTable,
...
+              if (staleSegmentsId.contains(segmentId)) {
+                segmentMetadataFile.delete();
               }
             }
           }
+        } else {
+          LOGGER.error(
+              "Not able to acquire the Table status lock for partial load deletion for table "
+                  + carbonTable.getDatabaseName() + "." + carbonTable.getTableName()
+                  + ", retry compaction");
```
Comment (akashrn5) [inline processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java:144]: here the error message is same as below, like retry compaction. Please check and change it.

Comment (ajantha-bhat) [reply processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java:144]: assigned to a variable

File: processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java (hunk 10 above)
Comment (ajantha-bhat) [reply processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java:129]: ok

File: core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java [hunk 12]
```diff
@@ -251,9 +261,29 @@ public String writeMergeIndexFileBasedOnSegmentFile(String segmentId,
     String path = CarbonTablePath.getSegmentFilesLocation(table.getTablePath())
         + CarbonCommonConstants.FILE_SEPARATOR + newSegmentFileName;
     if (!table.isHivePartitionTable()) {
-      SegmentFileStore.writeSegmentFile(segmentFileStore.getSegmentFile(), path);
-      SegmentFileStore.updateTableStatusFile(table, segmentId, newSegmentFileName,
+      String content = SegmentStatusManager.readFileAsString(path);
+      try {
+        SegmentFileStore.writeSegmentFile(segmentFileStore.getSegmentFile(), path);
+      } catch (Exception ex) {
+        // delete merge index file if created,
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java:268]: In both the callers, they are catching these exceptions and adding log. Same behavior.

File: core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java [hunk 13]
```diff
@@ -233,15 +234,24 @@ public String writeMergeIndexFileBasedOnSegmentFile(String segmentId,
...
         for (PartitionSpec partitionSpec : partitionSpecs) {
           if (partitionSpec.getLocation().toString().equals(partitionPath)) {
-            SegmentFileStore.writeSegmentFile(table.getTablePath(), mergeIndexFile, partitionPath,
-                segmentId + "_" + uuid + "", partitionSpec.getPartitions(), true);
+            try {
+              SegmentFileStore.writeSegmentFile(table.getTablePath(), mergeIndexFile, partitionPath,
+                  segmentId + "_" + uuid + "", partitionSpec.getPartitions(), true);
+            } catch (Exception ex) {
+              // delete merge index file if created,
+              // keep only index files as segment file writing is failed
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java:251]: In both the callers, they are catching this exceptions and adding log. Same behavior.

File: core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java [hunk 14]
```diff
@@ -233,15 +234,24 @@ public String writeMergeIndexFileBasedOnSegmentFile(String segmentId,
...
         }
       }
       if (table.isHivePartitionTable()) {
         for (PartitionSpec partitionSpec : partitionSpecs) {
           if (partitionSpec.getLocation().toString().equals(partitionPath)) {
-            SegmentFileStore.writeSegmentFile(table.getTablePath(), mergeIndexFile, partitionPath,
-                segmentId + "_" + uuid + "", partitionSpec.getPartitions(), true);
+            try {
+              SegmentFileStore.writeSegmentFile(table.getTablePath(), mergeIndexFile, partitionPath,
+                  segmentId + "_" + uuid + "", partitionSpec.getPartitions(), true);
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java:248]: > name the boolean variable
This is java. Handled only _ change.

File: core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (akashrn5) [reply core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java:269]: what i meant to say is, here we can error as writing segment file failed, and then throw exception, because we just get the IO exception here and not any custom message exception, so if we add here, it will be easy for any future analysis.

File: processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java [hunk 15]
```diff
@@ -64,45 +68,85 @@ public static void deletePartialLoadDataIfExist(CarbonTable carbonTable,
...
+              staleSegmentMetadataFile.delete();
+            }
+          }
+        } else {
+          String errorMessage =
+              "Not able to acquire the Table status lock for partial load deletion for table "
+                  + carbonTable.getDatabaseName() + "." + carbonTable.getTableName()
+                  + ", retry compaction";
+          LOGGER.error(errorMessage);
+          if (isCompactionFlow) {
```
Comment (akashrn5) [inline processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java:144]: here its little bit confusing for me, the message is same, saying retry compaction, then why throw inside `isCompactionFlow` condition. If we want to throw inside the if condition, then the error message should be different when not compaction flow  right?

Comment (ajantha-bhat) [reply processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java:144]: ok, now the comment bit clear. I will change it

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2337/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4077/

File: core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java:252]: ok

File: core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java:271]: ok

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4084/

**Comment by ajantha-bhat** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4085/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2345/

**Comment by akashrn5** [comment]
LGTM
